<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Demo - SCISSOR</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "Demo";
    var mkdocs_page_input_path = "example.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> SCISSOR</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../installation/">Getting Started</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../tutorial/">Tutorial</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">Demo</a>
    <ul class="current">
    </ul>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">SCISSOR</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Demo</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <p>A toy dataset is available in the <code>SCISSOR</code> package:</p>
<pre><code class="language-r">library(SCISSOR)
data(TOY_data)
</code></pre>
<p>This dataset contains base-level coverage of a hypothetical gene <em>TOY</em> including three exons and the whole intronic regions between the exons. Among n=260 cases, there are eight samples with strong shape changes. Here, we aim to identify these eight shape changes using <code>SCISSOR</code>. </p>
<h2 id="data-preparation">Data preparation</h2>
<p>We first obtain pileup data matrix using <code>build_pileup</code> and genomic ranges using <code>get_Ranges</code> that will be used as inputs for the function <code>Scissor</code>: </p>
<pre><code class="language-r">Gene = toygene_symbol
regions = toygene_regions
pileupData = build_pileup(Pileup=toygene_coverage,regions=regions,
                          inputType=&quot;whole_intron&quot;,outputType=&quot;part_intron&quot;)
geneRanges = get_Ranges(Gene=&quot;TOY&quot;,regions=regions,outputType=&quot;part_intron&quot;)
caseIDs = colnames(pileupData)
</code></pre>
<p>In practice, you will often obtain pileup data directly from BAM files. This can be done with <code>read_BAM</code> (see <a href="../tutorial/#get-coverage-from-bam-files">Tutorial</a>). If you do not specify sample IDs, the <code>build_pileup</code> and <code>read_BAM</code> will generate the IDs by default as follows:</p>
<pre><code class="language-r">&gt; head(caseIDs)
[1] &quot;case-1&quot; &quot;case-2&quot; &quot;case-3&quot; &quot;case-4&quot; &quot;case-5&quot; &quot;case-6&quot;
</code></pre>
<h2 id="statistical-analysis">Statistical analysis</h2>
<p>The whole procedures of SCISSOR can be performed by using all-in-one function <code>Scissor</code> (see <a href="../tutorial/#run-scissor">Tutorial</a>). </p>
<pre><code class="language-r">ScissorOutput=Scissor(pileupData=pileupData,Ranges=geneRanges)
</code></pre>
<p>The object <strong>ScissorOutput</strong> is a large list containing the results. The identified shape changes are saved in <code>ScissorOutput$SC</code> and the resulting outlyingness scores from each of the global and local procedures are saved in <code>ScissorOutput$GSC$OS</code> and <code>ScissorOutput$LSC$OS</code>, respectively. You can plot the outlyingness scores using <code>plot_ScissorOS</code>:</p>
<pre><code class="language-r">par(mfrow=c(1,2))
plot_ScissorOS(object=ScissorOutput,colmat=NULL,textSC=TRUE)
</code></pre>
<p><img alt="figure_OS" src="../images/TOY_OS_kdeplot.png" /></p>
<p>The figures show the indices of outlier samples with the kernel desntify estimates from the scores. </p>
<p>You can also visualize the identified shape changes using <code>plot_pileup</code>:</p>
<pre><code class="language-r">par(mfrow=c(4,2),mar=c(2,2,2,2))
for (case in ScissorOutput$SC) {
  plot_pileup(Pileup=ScissorOutput$logData,Ranges=geneRanges,cases=case,
              logcount=ScissorOutput$logshiftVal)
}
</code></pre>
<p><img alt="figure_SC" src="../images/TOY_results.png" /></p>
<p>The first seven outliers are associated with exon skipping or intron retention whereas the last outlier is showing small deletion in the middle of the first exon. </p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="../tutorial/" class="btn btn-neutral" title="Tutorial"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../tutorial/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
